### PR TITLE
Filter Agent events by tag

### LINKED ISSUE #3396: Filter events by tag
## Description

Add the ability to filter events by tag, similarly to how they can be filtered by type.

## Note
The current filtering approach is a simple solution that covers many, but not all use cases. Consider switching to GraphQL for advanced searching instead of using one of many poorly-defined "REST" approaches.

## Tasks
- [x] Add the capability to filter events by tag (0d)

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes #3396

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [ ] ~Was the documentation framework updated to reflect the changes?~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [ ] ~Do all end-to-end tests pass?~
* [x] Any other testing performed?
    > Tested by running Island and querying endpoints manually
* [ ] ~If applicable, add screenshots or log transcripts of the feature working~

### CODE REVIEW

File: monkey/monkey_island/cc/resources/agent_events.py [hunk 1]
```diff
@@ -68,16 +69,39 @@ def get(self):
...
             type_ = None if type_arg is None else self._agent_event_registry[type_arg]
         except KeyError:
-            raise Exception(f'Unknown agent event type "{type_}"')
+            raise Exception(f'Unknown agent event type "{type_arg}"')
+
+        return type_
+
+    def _parse_tag_arg(self, tag_arg: Optional[str]) -> Optional[str]:
+        if tag_arg and not re.match(pattern=re.compile(EVENT_TAG_REGEX), string=tag_arg):
+            raise Exception(f'Invalid event tag "{tag_arg}"')
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/resources/agent_events.py:100]: Errors should be specific.
```suggestion
            raise ValueError(f'Invalid event tag "{tag_arg}"')
```

File: monkey/monkey_island/cc/resources/agent_events.py [hunk 2]
```diff
@@ -68,16 +69,39 @@ def get(self):
...
+        timestamp_constraint = self._parse_timestamp_arg(timestamp_arg)
+
+        return type_, tag, success, timestamp_constraint
+
+    def _parse_type_arg(self, type_arg: Optional[str]) -> Optional[Type[AbstractAgentEvent]]:
         try:
             type_ = None if type_arg is None else self._agent_event_registry[type_arg]
         except KeyError:
-            raise Exception(f'Unknown agent event type "{type_}"')
+            raise Exception(f'Unknown agent event type "{type_arg}"')
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/resources/agent_events.py:94]: Also ValueError.

File: monkey/monkey_island/cc/resources/agent_events.py [hunk 3]
```diff
@@ -106,20 +133,32 @@ def _parse_event_filter_args(
...
+
+        if tag is not None:
+            events_by_tag: Sequence[
+                AbstractAgentEvent
+            ] = self._agent_event_repository.get_events_by_tag(tag)
+        else:
+            events_by_tag = self._agent_event_repository.get_events()
+
+        # this has better time complexity than converting both lists to sets,
+        # finding their intersection, and then sorting the resultant set by timestamp
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/resources/agent_events.py:160]: Good work.

File: monkey/monkey_island/cc/resources/agent_events.py [hunk 4]
```diff
@@ -106,20 +133,32 @@ def _parse_event_filter_args(
...
+                AbstractAgentEvent
+            ] = self._agent_event_repository.get_events_by_type(type_)
         else:
-            events = self._agent_event_repository.get_events()
+            events_by_type = self._agent_event_repository.get_events()
+
+        if tag is not None:
+            events_by_tag: Sequence[
+                AbstractAgentEvent
+            ] = self._agent_event_repository.get_events_by_tag(tag)
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/resources/agent_events.py:155]: A minor performance improvement might be to extract this as a method and just say `return events_by_type`

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/resources/agent_events.py (hunk 4 above)
Comment (shreyamalviya) [reply monkey/monkey_island/cc/resources/agent_events.py:155]: I don't think I understand. Extract what exactly?

**Pull request merged**